### PR TITLE
Fix sliding window attn mask

### PR DESCRIPTION
**Pull request by remi-or**
It was reporedt in #41184 that the sliding window attention mask has errors in CB. This PR fixes this and adds a test for the `build_attention_mask` function, so it does not happen again.

### CODE REVIEW

**Comment by remi-or** [comment]
@bot /style

File: src/transformers/generation/continuous_batching/continuous_api.py [hunk 1]
```diff
@@ -63,8 +63,8 @@ def build_attention_mask(
         masked = torch.triu(minus_inf, diagonal=causal_diagonal)
         # Apply sliding window mask if needed
         if sliding_window > 1:
-            sliding_diagonal = seqlen_k - seqlen_q + sliding_window
-            masked = torch.tril(masked, diagonal=sliding_diagonal)
+            sliding_diagonal = seqlen_k - seqlen_q - sliding_window
+            masked += torch.tril(minus_inf, diagonal=sliding_diagonal)
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/continuous_api.py:116]: Mmmm the sliding indeed has to be substracted but only if seqlen_q > sliding

**Review by ArthurZucker — COMMENTED** [review]
good that we are having a look. 
If we have:
```
k = 200
q = 200
sliding = 496
k - q - 496 = -496 
```
Probably noop? but not sure 
```
k = 800 
q = 1
sliding = 200
k - q - sliding = 599
```
the 600 index will be attended to, this does look correct

``` 
k = 200 
q = 1
sliding = 400
k - sliding - q = -201
```
not sur that's what we want?

File: src/transformers/generation/continuous_batching/continuous_api.py (hunk 1 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/continuous_api.py:116]: I am not sure of this, because the formula works with all those configs: 
```
k = 10, q = 10, sliding = 20
k = 10, q = 1, sliding = 4
k = 10, q = 1, sliding = 20
```

Comment (remi-or) [reply src/transformers/generation/continuous_batching/continuous_api.py:116]: Added ASCII drawing, should make things clearer

**Review by ArthurZucker — APPROVED** [review]
thanks a lot easier to read
